### DISCUSSION TITLE
github.sha isn't the value expected

### DISCUSSION THREAD

**Discussion by anderspitman** [community/community#25191] [Actions]
<p>I’m trying to retrieve the SHA from the commit that triggers the action. So if I make a commit and then do a push, within that push action handler I want to get that commit SHA. My goal is to build a URL based around the SHA and then comment on the PR with the URL (using <a href="https://github.com/unsplash/comment-on-pr" rel="nofollow noopener">https://github.com/unsplash/comment-on-pr</a>).</p>
<p>I expected github.sha to work, but I have no idea where the values it’s giving me are coming from. They don’t match any of the commits in the repo. Here’s the code I tried:</p>
<pre><code>name: CI

on:
  push:
    branches: [master]
  pull_request:
    branches: [master]
jobs:
  example:
    name: sample comment
    runs-on: ubuntu-latest
    steps:
      - name: comment PR
        uses: unsplash/comment-on-pr@master
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          msg: "example.com/${{github.sha}}"
</code></pre>

**Comment by Yanjingzhu** [comment, marked as answer]
<p><a class="mention" href="/anderspitman">@anderspitman</a> When a workflow is triggered by pull request event, the github.ref will be refs/pull/&lt;id&gt;/merge. And the github.sha will be the pre merge commit id. If you have a pull request from <strong>dev</strong> to <strong>master,</strong> github.sha is not the commit which you pushed to dev.  </p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/ed2ecc670926443882baf2982432846b12d84db8.png" data-download-href="/uploads/short-url/xQdqopVETsJULCWrNlP9RB659Gg.png?dl=1" title="pr event.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181089328-ae172884-e962-4439-8742-9c369791e10a.png" alt="pr event.png" data-base62-sha1="xQdqopVETsJULCWrNlP9RB659Gg" width="549" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/ed2ecc670926443882baf2982432846b12d84db8_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">pr event.png</span><span class="informations">768×699 29.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>You need to use <strong>github.event.pull_request.head.sha</strong> to get the commit id which you pushed to dev. </p>
<p>And based on my test when the workflow is triggered by push to master,  <strong>comment PR</strong> step will fail. </p>
<p>So you could change your workflow file to </p>
<pre><code>name: CI
on:
  pull_request:
    branches: [master]

jobs:
  example:
    name: sample comment
    runs-on: ubuntu-latest
    steps:
      - name: comment PR
        uses: unsplash/comment-on-pr@master
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          msg: "example.com/${{github.event.pull_request.head.sha}}"</code></pre>

**Reply by wheelerlaw** [reply]
Is there an explanation for this counter-intuitive behavior? It really seems like poor design to have `github.ref` refer to anything but the head of the feature branch.

**Comment by anderspitman** [comment]
<p>Sorry for the late response. This does indeed do the trick. Thanks!</p>

**Comment by mongodben** [comment]
i recently encountered the same issue, but `github.event.pull_request.head.sha` was not working for me.

what worked was using `github.event.workflow_run.head_sha`. hope this helps!

**Reply by RomeroGaliza** [reply]
depends on the event you're dealing with: `pull_request` vs. `workflow_run`. ;)
